File: src/main/java/com/github/dkoval/leetcode/challenge/ConvertSortedArrayToBinarySearchTree.java
package com.github.dkoval.leetcode.challenge;

import com.github.dkoval.leetcode.TreeNode;

/**
 * <a href="https://leetcode.com/explore/challenge/card/july-leetcoding-challenge-2021/611/week-4-july-22nd-july-28th/3827/">Convert Sorted Array to Binary Search Tree</a>
 * <p>
 * Given an integer array nums where the elements are sorted in ascending order, convert it to a height-balanced binary search tree.
 * <p>
 * A height-balanced binary tree is a binary tree in which the depth of the two subtrees of every node never differs by more than one.
 */
public class ConvertSortedArrayToBinarySearchTree {

    // O(N) time | O(N) space for the call stack
    public TreeNode sortedArrayToBST(int[] nums) {
        return buildBST(nums, 0, nums.length - 1);
    }

    private TreeNode buildBST(int[] nums, int start, int end) {
        if (start > end) {
            return null;
        }
        int mid = start + (end - start) / 2;
        TreeNode root = new TreeNode(nums[mid]);
        root.left = buildBST(nums, start, mid - 1);
        root.right = buildBST(nums, mid + 1, end);
        return root;
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/interview/trees/ConvertSortedArrayToBinarySearchTree.kt
package com.github.dkoval.leetcode.interview.trees

import com.github.dkoval.leetcode.TreeNode

/**
 * [Convert Sorted Array to Binary Search Tree](https://leetcode.com/explore/featured/card/top-interview-questions-easy/94/trees/631/)
 *
 * Given an array where elements are sorted in ascending order, convert it to a height balanced BST.
 *
 * For this problem, a height-balanced binary tree is defined as a binary tree in which the depth of
 * the two subtrees of every node never differs by more than 1.
 */
object ConvertSortedArrayToBinarySearchTree {

    fun sortedArrayToBST(nums: IntArray): TreeNode? {
        if (nums.isEmpty()) return null
        return doSortedArrayToBST(nums, 0, nums.lastIndex)
    }

    private fun doSortedArrayToBST(nums: IntArray, startIndex: Int, endIndex: Int): TreeNode? {
        if (startIndex > endIndex) return null
        val midIndex = startIndex + (endIndex - startIndex) / 2
        val curr = TreeNode(nums[midIndex])
        curr.left = doSortedArrayToBST(nums, startIndex, midIndex - 1)
        curr.right = doSortedArrayToBST(nums, midIndex + 1, endIndex)
        return curr
    }
}